io/zenoh-links/zenoh-link-shm/src/unix/unicast.rs

```diff
@@ -0,0 +1,573 @@
+        }
+
+        let suffix_bytes: &[u8; 4] = &msg[4..].try_into()?;
+        let suffix = u32::from_ne_bytes(*suffix_bytes);
+        Ok(suffix)
+    }
+
+    async fn confirm(suffix: u32, pipe: &mut PipeW) -> ZResult<()> {
+        Self::send(suffix, pipe).await
+    }
+
+    async fn expect(expected_suffix: u32, pipe: &mut PipeR) -> ZResult<()> {
+        let recived_suffix = Self::receive(pipe).await?;
+        if recived_suffix != expected_suffix {
+            bail!(
+                "Suffix mismatch: expected {} gor {}",
+                expected_suffix,
+                recived_suffix
+            )
+        }
+        Ok(())
+    }
+}
+
+struct PipeR {
+    pipe: Async<File>,
+}
+
+impl Drop for PipeR {
+    fn drop(&mut self) {
+        if let Ok(path) = self.pipe.as_mut().path() {
+            let _ = unlink(&path);
+        }
+    }
+}
+impl PipeR {
+    async fn new(path: &str, access_mode: u32) -> ZResult<Self> {
+        // create, open and lock named pipe
+        let pipe_file = Self::create_and_open_unique_pipe_for_read(path, access_mode).await?;
+        // create async_io wrapper for pipe's file descriptor
+        let pipe = Async::new(pipe_file)?;
+        Ok(Self { pipe })
+    }
+
+    async fn read<'a>(&'a mut self, buf: &'a mut [u8]) -> ZResult<usize> {
+        let result = self
+            .pipe
+            .read_with_mut(|pipe| match pipe.read(&mut buf[..]) {
+                Ok(0) => Err(async_std::io::ErrorKind::WouldBlock.into()),
+                Ok(val) => Ok(val),
+                Err(e) => Err(e),
+            })
+            .await?;
+        ZResult::Ok(result)
+    }
+
+    async fn read_exact<'a>(&'a mut self, buf: &'a mut [u8]) -> ZResult<()> {
+        let mut r: usize = 0;
+        self.pipe
+            .read_with_mut(|pipe| match pipe.read(&mut buf[r..]) {
+                Ok(0) => Err(async_std::io::ErrorKind::WouldBlock.into()),
+                Ok(val) => {
+                    r += val;
+                    if r == buf.len() {
+                        return Ok(());
+                    }
+                    Err(async_std::io::ErrorKind::WouldBlock.into())
+                }
+                Err(e) => Err(e),
+            })
+            .await?;
+        ZResult::Ok(())
+    }
+
+    async fn create_and_open_unique_pipe_for_read(path_r: &str, access_mode: u32) -> ZResult<File> {
+        let r_was_created = create(path_r, Some(access_mode));
+        let open_result = Self::open_unique_pipe_for_read(path_r);
+        match (open_result.is_ok(), r_was_created.is_ok()) {
+            (false, true) => {
+                // clean-up in case of failure
+                let _ = remove_file(path_r).await;
+            }
+            (true, false) => {
+                // drop all the data from the pipe in case if it already exists
+                let mut buf: [u8; 1] = [0; 1];
+                while let Ok(val) = open_result.as_ref().unwrap().read(&mut buf) {
```

Comment:
Remove the `unwrap()` by replacing the `match { (bool, bool) .. }` with `match { (Ok(val), Err(_)) .. }`. I.e. matching directly on `Result` variants`.

Reply:
Yes, very good comment, I will get rid of unwrap through value decomposition as you recommended during the meetup